vshard/storage/init.lua

```diff
@@ -3743,20 +3744,44 @@ local function storage_cfg_build_local_box_cfg(cfgctx)
         box_cfg.read_only = not cfgctx.is_master_in_cfg
     end
     if type(box.cfg) == 'function' then
-        box_cfg.instance_uuid = box_cfg.instance_uuid or cfgctx.instance_uuid
-        box_cfg.replicaset_uuid = box_cfg.replicaset_uuid or cfgctx.replicaset_uuid
+        if cfgctx.new_cfg.identification_mode == 'name_as_key' then
+            box_cfg.instance_name = cfgctx.instance_id
+            box_cfg.replicaset_name = cfgctx.replicaset_id
+        else
+            box_cfg.instance_uuid = box_cfg.instance_uuid or
+                cfgctx.instance_id
+            box_cfg.replicaset_uuid = box_cfg.replicaset_uuid or
+                cfgctx.replicaset_id
+        end
     else
         local info = box.info
-        if cfgctx.instance_uuid ~= info.uuid then
-            error(string.format('Instance UUID mismatch: already set ' ..
-                                '"%s" but "%s" in arguments', info.uuid,
-                                cfgctx.instance_uuid))
-        end
-        local real_rs_uuid = util.replicaset_uuid(info)
-        if cfgctx.replicaset_uuid ~= real_rs_uuid then
-            error(string.format('Replicaset UUID mismatch: already set ' ..
-                                '"%s" but "%s" in vshard config',
-                                real_rs_uuid, cfgctx.replicaset_uuid))
+        if cfgctx.new_cfg.identification_mode == 'name_as_key' then
+            -- Strictly verify names in case of reconfiguration. There's no
+            -- other way to check, that config is applied to the correct
+            -- instance. Changing and applying names must be done externally.
```

Comment:
Not necessarily externally. If the names are used from the beginning, then vshard will set them on bootstrap in the code block above.

Reply:
Yeah, fixed the commit message